src/partisan_util.erl

```diff
@@ -131,7 +122,68 @@ maybe_connect_listen_addr(Node, ListenAddr, Connections0) ->
     Connections.
 
 %% @private
--spec connect(Node :: node_spec(), listen_addr()) -> {ok, pid()} | ignore | {error, term()}.
-connect(Node, ListenAddr) ->
+-spec connect(Node :: node_spec(), listen_addr(), channel()) -> {ok, pid()} | ignore | {error, term()}.
+connect(Node, ListenAddr, Channel) ->
     Self = self(),
-    partisan_peer_service_client:start_link(Node, ListenAddr, Self).
+    partisan_peer_service_client:start_link(Node, ListenAddr, Channel, Self).
+
+%% @doc Return a pid to use for message dispatch.
+dispatch_pid(Entries) ->
+    dispatch_pid(undefined, Entries).
+
+%% @doc Return a pid to use for message dispatch for a given channel.
+dispatch_pid(Channel, Entries) ->
+    %% Entries for channel.
+    ChannelEntries = lists:filter(fun({_, C, _}) ->
+        case C of
+            Channel ->
+                true;
+            _ ->
+                false
+        end
+    end, Entries),
+
+    %% Fall back to unlabeled channels.
+    DispatchEntries = case length(ChannelEntries) of
+        0 ->
+            Entries;
+        _ ->
+            ChannelEntries
+    end,
+
+    %% Randomly select one.
+    {_ListenAddr, _Channel, Pid} = lists:nth(rand_compat:uniform(length(DispatchEntries)), DispatchEntries),
+
+    Pid.
+
+%% @private
+maybe_initiate_parallel_connections(Connections0, Channel, Node, ListenAddr, Parallelism, Pids) ->
+    FilteredPids = lists:filter(fun({A, C, _}) ->
+                            case A of
+                                ListenAddr ->
+                                    case C of
+                                        Channel ->
+                                            true;
+                                        _ ->
+                                            false
+                                    end;
+                                _ ->
+                                    false
+                            end
+                    end, Pids),
+    case length(FilteredPids) < Parallelism andalso Parallelism =/= undefined of
```

Comment:
Do you need to do any of this work if `Parallelism == undefined`? If not, you could skip it by pattern-matching on that in the function head.